IECoreScenePreview::Renderer: allow declaring unprefixed custom options

mattigruener opened the pull request:
@johnhaddon, I guess we might also want to consider doing this for attributes?

I added a test for the option's value, but I don't check the warning yet... Let me know if you think it's necessary to add that.

johnhaddon commented on python/GafferArnoldTest/IECoreArnoldPreviewTest/RendererTest.py:
```diff
@@ -1435,6 +1435,26 @@ def testStepSizeAttributeDefersToProceduralParameter( self ) :
+		r.render()
+		del r
+
+		with IECoreArnold.UniverseBlock() :
```
Should be `with IECoreArnold.UniverseBlock( writable = True ) :`, since loading the ass file edits the universe.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1439,6 +1439,18 @@ class ArnoldRenderer : public IECoreScenePreview::Renderer
+				}
+				else
+				{
+					ParameterAlgo::setParameter( options, name.c_str() + 11, value );
```
Bear in mind that the value may be NULL - this is used in interactive rendering to remove a previous option that is no longer present. So I think you need the same here as there is in the "ai:" case below - using `AiNodeResetParameter()` when value is NULL. You might want to fire up an IPR session to check it works though (don't worry about a unit test for this bit, it'd be a real pain to set up).

mattigruener commented:
Hey John,

I updated the PR. Addressing your comment regarding the test was easy enough, but I'm not super happy with the other one.

I noticed that updating the parameter was resulting in a warning that the parameter already exists. That happened because `ParameterAlgo::setParameter` isn't clever enough to not redeclare an existing user parameter - it only looks at the `AtNodeEntry` and not at the parameters that exist on the node instance. I work around it here by checking the instance, throwing away the parameter and recreating it. If NULL is passed in as value we just don't recreate it.

I had a bit of a hard time coming up with a proper test for this. No user option affects the actual rendering at the moment, so it's hard to see whether the update was done correctly. I tried using a simple OSL shader to fetch the option, but Arnold doesn't seem to support getattr() for user options yet. The only option that would affect the rendering is the Cryptomatte one that I mentioned before, but it seems that changing that option does not trigger a reevaluation of the outputs that were already created? Either way - I tried to test it by putting prints into the code that query the options through the Arnold API after updating the parameter. That seems to work fine and I decided to move on for today.

Are you happy with the implementation and the level of testing or do you want me to look at it again?

Thanks,
Matti.

johnhaddon commented:
Looks good - thanks Matti.
